rfd/0228-resource-scoped-constraints-in-access-requests.md

```diff
@@ -0,0 +1,168 @@
+    SshConstraints ssh = 11;
+    DbConstraints  db  = 12;
+  }
+}
+
+message AwsConstraints {
+  repeated string role_arns       = 1; // AWS Console
+  repeated string permission_sets = 2; // AWS IAM IC
+}
+
+message SshConstraints {
+  repeated string logins = 1;
+  bool allow_sudo        = 2;        // example non-principal constraint
+}
+
+message DbConstraints {
+  repeated string roles = 1;
+}
+```
+
+### RBAC Semantics with Constraints
+- **Validation & resolution**: Requests that include `constraints` are validated using constraint-aware role matchers derived from each `ResourceID.Constraints` (e.g., ARN/permission set/SSH login matchers). The validator extends the existing `applicableSearchAsRoles`/`roleAllowsResource` flow by passing these matchers, so a role only qualifies if it allows the requested resource *and* the specified constraints. Resources are still pre-filtered by current access and  requested `ResourceIDs` are still deduplicated and size-capped, as they are currently. The resulting role list is the set of requestable/search-as roles that satisfies all `(resource, constraints)` pairs (when multiple roles qualify, behavior matches current semantics). If no roles qualify, existing failure behavior and errors apply.
+- **Authorization scope**: On approval/assumption, certs carry the exact `(resource, constraints)` pairs via `tlsca.Identity.AllowedResourceIDs` (each entry including `ResourceID.Constraints`). Access checking enforces those constraints: if a role grants multiple principals on a resource, approving a request for a narrower set (e.g., a single ARN, specific SSH logins, or IC permission sets) must only allow those requested items. By reading `ResourceID.Constraints` from `AllowedResourceIDs`, the Access Checker scopes evaluation accordingly and denies any un-requested principals even if the resolved role(s) would otherwise permit them.
+
+### Web/API (high-level)
+- Listing UnifiedResources surfaces per-constraint status when "show requestable" is enabled. For each item (ARN, permission set, SSH login), a `requiresRequest: true|false` marker is provided so the UI can render a single menu with "Granted" and "Requestable" sections.
+- Access Requests that include `ResourceID.constraints` present both requested constraints and resolved roles to reviewers.
+
+## Compatibility
+- The `constraints` field is optional; existing clients and requests without constraints continue to work unchanged.
+- UI can adopt the per-constraint "granted vs requestable" grouping incrementally by resource type.
+- RBAC semantics are additive. Constraint matchers are considered only when constraints are present in a request.
+
+## Out of Scope
+- Final UI designs and UX specifics (layout, grouping visuals, request checkout/review details).
+- `tsh request` commands and Teleport Connect flows.
+- Exhaustive resource/constraint coverage beyond the initial AWS Console/AWS IC focus (SSH/DB shapes and user stories provided as examples).
```

Comment:
So this entire section is not really out of scope, I would reframe this into a staged delivery plan/process: phase 1 is X, phase is Y, phase 3 is Connect support, etc.